Add bucksense Adapter

bkse-stefanodechicchis opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [X ] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 

## Description of change
This is first release for Bucksense Adapter

<!-- For new bidder adapters, please provide the following -->
- test parameters for validating bids
```
{
  bidder: 'bucksense',
  params: {
    placementId : 1000
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer (stefano.dechicchis@bucksense.com)
- [ X] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

## Other information
<!-- References to related PR or issue #s, @mentions of the person or team responsible for reviewing changes, etc. -->

bretg commented:
docs PR https://github.com/prebid/prebid.github.io/pull/1284

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+   * @return boolean True if this is a valid bid, and false otherwise.
+  */
+  isBidRequestValid: function (bid) {
+    if (isNaN(bid)) {
```
The `bid` param is an object at this stage of the code, so this if statement check will always return `true` and cause the bidRequest for your bidder to be considered invalid.

Can you review this logic and adjust it accordingly?

bkse-stefanodechicchis replied:
fixed with new Push

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+        data: {
+          'pub_id': location.host,
+          'pl_id': bid.params.placementId,
+          'sys_href': encodeURI(location.href),
```
If you haven't already looked at it, there are a set of data points you can use in your adapter to get information on the webpage loading the adapter.  Please see the link below and consider using them if they would fit your needs:
http://prebid.org/dev-docs/bidder-adaptor.html#referrers

bkse-stefanodechicchis replied:
Thanks for point to this but in this release we would like to keep using the location object if it is not a problem

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+  * @param {ServerResponse[]} serverResponses List of server's responses.
+  * @return {UserSync[]} The user syncs which should be dropped.
+  */
+  getUserSyncs: function(syncOptions, serverResponses) {
```
Is this code still a work-in-progress?

If not and you're not planning to support userSyncs (at least at this time) - you can just remove this function entirely from your adapter.

bkse-stefanodechicchis replied:
function not needed on this release, we removed.

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+   * @param ...
+   * @return ...
+  */
+  onTimeout: function(timeoutData) {
```
Similar to the getUserSyncs function - you can remove this function if you're not planning to use it.

bkse-stefanodechicchis replied:
function not needed on this release, we removed.

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+   * @param {object} bid The bid to validate.
+   * @return ...
+  */
+  onBidWon: function(bid) {
```
Similar to the getUserSyncs function - you can remove this function if you're not planning to use it.

bkse-stefanodechicchis replied:
function not needed on this release, we removed.

jsnellbaker commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+
+const WHOIS = 'BKSHBID-008';
+const BIDDER_CODE = 'bucksense';
+const URL = 'https://prebid.bksn.se:445/prebid';
```
Is this the correct endpoint URL for your adapter?

When I tried to test the adapter, the POST requests to this URL were resulting in a 301 Moved Permanently response.

bkse-stefanodechicchis replied:
Yes request is doing a 301 but then got redirect to the script that return the BID. I have tested the prebid on the Hello World example. Let me know if you do not see the response at all from your test.

bkse-stefanodechicchis commented on modules/bucksenseBidAdapter.js:
```diff
@@ -0,0 +1,171 @@
+import { registerBidder } from 'src/adapters/bidderFactory';
+import { BANNER } from 'src/mediaTypes';
```
fixed with new Push

jsnellbaker commented:
Hi @bkse-stefanodechicchis, Thanks for making the updates.

I retested the adapter and was now able to see the bid return and render successfully.

However, when I reran the coverage tests - the adapter is still below the ideal; it's at 72% currently.
Can you put together some additional unit tests to fill in the gaps?

Also can you update the various `console.log` statements in the adapter file to use the`utils.logInfo()` function instead?  This function would categorize/brand the messages in the console that they're part of Prebid and would be more ideal to use.

We should be good to merge once these last few items are addressed.  Thanks.

bkse-stefanodechicchis commented:
Hi,
I'm struggling to increase the percentage of testing.
It is my first experience with Chai and this is not helping.
My specs file already test all the 4 function that are in the Adapter, so I
thought I would add more scenarios to check in each function, but if I add
2 new scenario, I don't see the % going up.
I appreciate any suggestions on what I'm doing wrong
Thanks


On Wed, May 1, 2019 at 12:36 PM jsnellbaker <notifications@github.com>
wrote:

> Hi @bkse-stefanodechicchis <https://github.com/bkse-stefanodechicchis>,
> Thanks for making the updates.
>
> I retested the adapter and was now able to see the bid return and render
> successfully.
>
> However, when I reran the coverage tests - the adapter is still below the
> ideal; it's at 72% currently.
> Can you put together some additional unit tests to fill in the gaps?
>
> Also can you update the various console.log statements in the adapter
> file to use theutils.logInfo() function instead? This function would
> categorize/brand the messages in the console that they're part of Prebid
> and would be more ideal to use.
>
> We should be good to merge once these last few items are addressed. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prebid/Prebid.js/pull/3785#issuecomment-488335386>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL5U3WC7XWBUQVZF6T774GTPTHBKLANCNFSM4HIYVZAQ>
> .
>

bkse-stefanodechicchis commented:
I'm struggling to increase the percentage of testing.
![Screen Shot 2019-05-03 at 4 12 47 PM](https://user-images.githubusercontent.com/50023896/57163159-5e9be680-6dbe-11e9-93be-f5a0e40cab82.png)

My specs file already test all the 4 functions that are in the Adapter, so I thought I would add more scenarios to check in each function, but if II added some more and I don't see the % going up, so I think I'm in the wrong direction.
I appreciate any suggestions/help on what I'm doing wrong
Thanks

jsnellbaker commented:
Hi @bkse-stefanodechicchis 

Just to clarify, you are looking at your specific adapter's coverage rate correct?  It's inside the modules sub-folder in that initial screen.  See below for example:

<img width="1664" alt="Screen Shot 2019-05-06 at 8 25 44 AM" src="https://user-images.githubusercontent.com/31102355/57224846-97170c80-6fd8-11e9-8986-08404dbd3428.png">

From what I can see (currently with what has been committed), the outstanding areas all relate to the `bksdebug` param.  

If you setup a pair of tests where that field is enabled (one for `requestBids` and one for `interpretResponse`) and swap out the `console.log` with `utils.logInfo` - you could create a stub of the `utils.logInfo` function and run your checks against that stub (ie check for the `called` property is `true` for example).  These new tests should help fill the gap.

bkse-stefanodechicchis commented:
I just commit a new version, I removed the `console.log` and use `utils.logInfo`.
I'm not able to reach the 80% coverage I also removed the bksdebug param and added some more use case but the % just not moving up.

I look at the file from the `gulp view-coverage`
![image](https://user-images.githubusercontent.com/50023896/57257501-a07b9580-7027-11e9-9231-96c4b51a1e23.png)
and I'm try to figure out why some IF or ELSE get a flag

jsnellbaker commented:
Hi @bkse-stefanodechicchis I checked out your updates to the adapter test cases.  I'm seeing your adapter coverage at 94% now (see below):
<img width="1656" alt="Screen Shot 2019-05-07 at 10 15 17 AM" src="https://user-images.githubusercontent.com/31102355/57306660-37406480-70b1-11e9-9962-a393aaf61bf6.png">

Are you not seeing the same thing?  If you're happy with how your adapter is handling certain scenarios (this is in reference to your last comment on the if/else statements in the `interpretResponse`), we should be okay to merge the PR.  

Please let me know when you have the chance.

bkse-stefanodechicchis commented:
Hi, I get confuse I was looking at column "Branches" and the % was low there.
Anyway we are good in our side, so ready for the merge from our point of view.
Thanks
